alternator/streams.cc

```diff
@@ -479,50 +494,125 @@ future<executor::request_return_type> executor::describe_stream(client_state& cl
+            prev = std::prev(i);
+            std::stable_sort(prev->second.streams.begin(), prev->second.streams.end(), [](auto& id1, auto& id2) { 
+                return id1.token() < id2.token(); 
+            });
+        }
+
         for (; limit > 0 && i != e; prev = i, ++i) {
             auto& [ts, sv] = *i;
 
             last = std::nullopt;
 
-            for (auto& id : sv.streams()) {
-                if (shard_start && shard_start->id != id) {
-                    continue;
-                }
-                if (shard_start && shard_start->id == id) {
-                    shard_start = std::nullopt;
-                    continue;
+            auto get_expiration = [&](auto iter) {
+                if (++iter != e) {
+                    return iter->first;
                 }
+                return db_clock::time_point{};
+            };
+
+            auto lo = sv.streams.begin();
+            auto end = sv.streams.end();
+
+            // first sort by index.
+            std::stable_sort(lo, end, [](auto& id1, auto& id2) {
+                return id1.index() < id2.index();
+            });
+
+            if (shard_start) {
+                // find next shard position
+                // first id set is the wraparound range. skip it here.
+                lo = std::find_if(lo, end, [](const cdc::stream_id& id) { return id.index() != 0; });
+                lo = std::upper_bound(lo, end, shard_start->hi_token, [](const dht::token& t, const cdc::stream_id& id) {
+                    return t < id.token();
+                });
+                shard_start = std::nullopt;
+            }
+
+            for (auto hi = lo; lo != end; lo = hi) {
+                auto& id = *lo++;
+
+                hi = std::find_if(lo, end, [i = id.index()](const cdc::stream_id& id) {
+                    return id.index() != i;
+                });
+                auto t = id.token();
+                auto et = std::prev(hi)->token();
+
+                last.emplace(ts, get_expiration(i), t, et);
 
                 auto shard = rjson::empty_object();
 
                 if (prev != e) {
-                    auto token = id.token();
-                    auto& pids = prev->second.streams();
-                    auto pid = std::upper_bound(pids.begin(), pids.end(), token, [](const dht::token& t, const cdc::stream_id& id) {
+                    auto& pids = prev->second.streams;
+                    // maybe re-sort by token, so we can find the previous shard bounds
+                    // if prev is sorted by index, we can just rotate it slightly to make it sorted by
+                    // token.
+                    if (pids.front().token() > pids.back().token()) {
+                        auto nf = std::find_if(pids.begin(), pids.end(), [t = pids.back().token()](const cdc::stream_id& id) {
+                            return id.token() < t;
+                        });
+                        std::rotate(pids.begin(), nf, pids.end());
```

Comment:
Maybe using a single sort which compares by index first and token second in the first place would be simpler than all of this.

But wouldn't sorting by just the token automatically result in sorting by index as well?

Reply:
Then we get zero == range that covers the rest of the id:s (being the wraparound). Could add code that "finds" the wraparound, but it seems rather like false abstraction to me. And a pointless search of a rather large array once the cluster becomes larger.
I think we can just define it as "index 0 == wraparound". Easier.